File: leetCode/Python/206-reverseLinkedList.py
# Reverse a singly linked list.

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def print_llst(head):
    llst = ""
    while head:
          llst += str(head.val)
          if head.next:
             llst += '->'
          head = head.next
    print llst

'''
Time:  O(n)
Space: O(1)
'''
class Solution(object): # iterative
    def reverse(self, head):
        start = None
        return self.reverse_helper(head, start) 

    def reverse_helper(self, head, start):
        if head is None:
            return start
        next = head.next
        head.next = start
        return self.reverse_helper(next, head)

'''
Time:  O(n)
Space: O(1)
'''
class Solution(object): # recursive
    def reverseList(self, head):
        return self.reverse_helper(head, None)

    def reverse_helper(self, head, start):
        if head is None:
            return start
        else:
            nxt = head.next
            head.next = start
            return self.reverse_helper(nxt, head)

if __name__ == '__main__':
   s = Solution()
   test = ListNode(1)
   test.next = ListNode(2)
   test.next.next = ListNode(3)
   test.next.next.next = ListNode(4)
   test.next.next.next.next = ListNode(5)

   print_llst(s.reverse(test))

File: leetCode/C++/206-ReverseLinkedList.cpp
#include <iostream>

using namespace std;

/**
 * Definition for singly-linked list.
 **/
struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
};

class Solution1 { // iterative
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *start = nullptr, *nxt = nullptr;
        while (head) {
             nxt = head->next;
             head->next = start;
             start = head;
             head = nxt;
        }
        return start;
    }
};

class Solution { // recursive
public:
    ListNode* reverseList(ListNode* head) {
        return reverse_llst_helper(head, nullptr);
    }
private:
    ListNode* reverse_llst_helper(ListNode* head, ListNode* start) {
        if (!head)
            return start;
        else {
            ListNode* nxt = head->next;
            head->next = start;
            return reverse_llst_helper(nxt, head);
        }
    }
};

void printLinkedList(ListNode *head) {
     string llst = "";
     while (head) {
            cout << head->val;
            if (head->next) {
                cout << " -> ";
            };
            head = head->next;
     };
     cout << endl;
}

int main(void) {
    Solution s;
    struct ListNode test(1);
    struct ListNode t1(2);
    struct ListNode t2(3);
    struct ListNode t3(4);
    struct ListNode t4(5);
    (&test)->next = &t1;
    (&test)->next->next = &t2;
    (&test)->next->next->next = &t3;
    (&test)->next->next->next->next = &t4;
    printLinkedList(s.reverseList(&test));
    return 0;
};